pkg/resource/secret_reconciler.go

```diff
@@ -68,53 +68,85 @@ func NewSecretPropagatingReconciler(m manager.Manager) reconcile.Reconciler {
 		ctx, cancel := context.WithTimeout(context.Background(), secretReconcileTimeout)
 		defer cancel()
 
-		// The 'from' secret is also know as the 'propagating' secret.
-		from := &corev1.Secret{}
-		if err := client.Get(ctx, req.NamespacedName, from); err != nil {
-			// There's no propagation to be done if the secret we're propagating
-			// from does not exist. We assume we have a watch on that secret and
-			// will be queued if/when it is created. Otherwise we'll be requeued
-			// implicitly because we return an error.
-			return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetSecret)
-		}
-
 		// The 'to' secret is also known as the 'propagated' secret. We guard
 		// against abusers of the propagation process by requiring that both
 		// secrets consent to propagation by specifying each other's UID. We
 		// cannot know the UID of a secret that doesn't exist, so the propagated
 		// secret must be created outside of the propagation process.
 		to := &corev1.Secret{}
-		n := types.NamespacedName{
-			Namespace: from.GetAnnotations()[AnnotationKeyPropagateToNamespace],
-			Name:      from.GetAnnotations()[AnnotationKeyPropagateToName],
-		}
-		if err := client.Get(ctx, n, to); err != nil {
+		if err := client.Get(ctx, req.NamespacedName, to); err != nil {
 			// There's no propagation to be done if the secret we propagate to
 			// does not exist. We assume we have a watch on that secret and will
 			// be queued if/when it is created. Otherwise we'll be requeued
 			// implicitly because we return an error.
 			return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetSecret)
 		}
 
-		if from.GetAnnotations()[AnnotationKeyPropagateToUID] != string(to.GetUID()) {
-			// The propagating secret expected a different propagated secret. We
+		toAnnotations := to.GetAnnotations()
+		var fromUID string
+		var fromName string
+		var fromNamespace string
+
+		for key, val := range toAnnotations {
+			// If this is a 'from' secret we do not want to reconcile and will
+			// instead defer to the enqueued request for the 'to' secret
+			if strings.HasPrefix(key, AnnotationKeyPropagateTo) {
+				return reconcile.Result{}, nil
+			}
+
+			if strings.HasPrefix(key, AnnotationKeyPropagateFrom) {
+				fromUID = strings.TrimPrefix(key, AnnotationKeyPropagateFrom)
+				f := strings.Split(val, "/")
+				if len(f) != 2 {
+					return reconcile.Result{}, errors.New(errInvalidPropagateFormat)
+				}
+				fromNamespace = f[0]
+				fromName = f[1]
+				break
+			}
+		}
```

Comment:
I think we could simplify this logic quite a lot if we used something similar to the contemporary `from` annotation, perhaps:

```
from.propagate.crossplane.io/uid: 8aedda70-2e77-11ea-978f-2e728ce88125
from.propagate.crossplane.io/namespace: default
from.propagate.crossplane.io/name: coolsecret 
```

Alternatively if we wanted to keep our annotations symmetrical (even though in practice we'd only support propagating from a single secret) then your suggestion of bailing out early if there is more than one from annotation would also allow us to simplify this block.